File: Problems/104. Maximum Depth of Binary Tree.cpp
#include<iostream>
#include<vector>

using namespace std;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 };


 
class Solution {
    public:

        void calculateDepthOfTree(TreeNode* node,int currentDepth){

                // check if this is a leaf node
                if(node->left == nullptr && node->right == nullptr){
                    // We mark this as depth of tree, only if it's greater
                    // than older depth
                    depth = max(depth,currentDepth);
                    return;
                }

                // continue dfs on left and right nodes 
                // (if they exist)

                if(node->left){
                    calculateDepthOfTree(node->left,currentDepth+1);
                }
                if(node->right){
                    calculateDepthOfTree(node->right,currentDepth+1);
                }

            }


        int maxDepth(TreeNode* root) {
            
            if(!root){
                return 0;
            }
            
            calculateDepthOfTree(root,1);
            return depth;
        }
    
    private:
        int depth = 1;
    
    
    
};

File: Problems/104. Maximum Depth of Binary Tree.java
// * Definition for a binary tree node.
class TreeNode {
      int val;
      TreeNode left;
      TreeNode right;
      TreeNode() {}
      TreeNode(int val) { this.val = val; }
      TreeNode(int val, TreeNode left, TreeNode right) {
          this.val = val;
          this.left = left;
          this.right = right;
      }
}

class TreeNodeAndDepth {
    TreeNode node;
    int depth;

    TreeNodeAndDepth(TreeNode node, int depth){
        this.depth = depth;
        this.node = node;
    }
}

class Solution {
    private int depthFirstTraversal(TreeNode node,int currentDepth){
        if(node==null || (node.left == null && node.right == null)){
            return  currentDepth;
        }
        return Math.max(depthFirstTraversal(node.left,currentDepth+1),depthFirstTraversal(node.right,currentDepth+1));
    }

    public int maxDepth(TreeNode root) {
        if(root==null){
            return 0;
        }
        return depthFirstTraversal(root,1);
    }
}
